Implement CE Subscriptions filters (#5715)
knative/eventing: docs/eventing-api.md

</tr>
<tr>
<td>
<code>filters</code><br/>
<em>
<a href="#eventing.knative.dev/v1.SubscriptionsAPIFilter">
[]SubscriptionsAPIFilter
</a>
</em>
</td>
<td>
<em>(Optional)</em>
<p>Filters is an experimental field that conforms to CNCF CloudEvents Subscriptions
API. It&rsquo;s An array of filter expressions that evaluates to true or false.
If any filter expression in the array evaluates to false, the event MUST
NOT be sent to the Subscriber. If all the filter expressions in the array
evaluates to true, the event MUST be attempted to be delivered. Absence of
a filter or empty array implies a value of true. In the event of users
specifying both Filter and Filters, then the later will override the first.
This will allow our users to try out the effect of the new filters field
without compromising existing Filter objects and try it out on existing

> Trigger objects.</p>

Reviewer: Allowing users to provide `Filter` and `Filters` might be counter-intuitive. I would expect a failure where a `one-of` is expected, besides I'm not sure how allowing both values will help trying it out (besides users having the previous `Filter` _at hand_)

Author: I agree and this is being added as an "experimental feature". I've tried to elaborate more on the reasoning behind this design in the [feature track](https://hackmd.io/@devguyio/H1Mg3my2d#FAQ) in the FAQ section. 

I believe that with the graduation of this experimental feature, the final flied will be `Filters` and the old `Filter` field will be deprecated. 

> I'm not sure how allowing both values will help trying it out (besides users having the previous Filter at hand)

The idea is exactly that, having the old filter at hand, and enabling the new filters language per trigger. Only during the experimental phase, as stated above.

</tr>
<tr>
<td>
<code>filters</code><br/>
<em>
<a href="#eventing.knative.dev/v1.SubscriptionsAPIFilter">
[]SubscriptionsAPIFilter
</a>
</em>
</td>
<td>
<em>(Optional)</em>
<p>Filters is an experimental field that conforms to CNCF CloudEvents Subscriptions

> API. It&rsquo;s An array of filter expressions that evaluates to true or false.

Reviewer 2: ```suggestion
API. It&rsquo;s an array of filter expressions that evaluate to true or false.
```

If any filter expression in the array evaluates to false, the event MUST
NOT be sent to the Subscriber. If all the filter expressions in the array

> evaluates to true, the event MUST be attempted to be delivered. Absence of

Reviewer 2: ```suggestion
evaluate to true, the event MUST be attempted to be delivered. Absence of
```

a filter or empty array implies a value of true. In the event of users

> specifying both Filter and Filters, then the later will override the first.

Reviewer 2: ```suggestion
specifying both Filter and Filters, then the latter will override the former.
```

> This will allow our users to try out the effect of the new filters field

Reviewer 2: (Not too comfortable with the idea of owning users in a OSS project, but that's subjective)
```suggestion
This will allow users to try out the effect of the new Filters field
```

without compromising existing Filter objects and try it out on existing

> Trigger objects.</p>

Reviewer 2: "object" appears twice in the same sentence, with 2 different meanings. Could we say "attribute-based Filter" instead of "Filter objects", or something along those lines?
